pkg/sql/schemachanger/scbuild/relation_common.go

```diff
@@ -49,66 +103,293 @@ func (b *buildContext) removeTypeBackRefDeps(
+		return
 	}
-	visitorDeleted := &tree.TypeCollectorVisitor{
-		OIDs: make(map[oid.Oid]struct{}),
+	defaultExpr := column.GetDefaultExpr()
+	sequenceIDs := make([]descpb.ID, 0, column.NumUsesSequences())
+	for idx := 0; idx < column.NumUsesSequences(); idx++ {
+		sequenceIDs = append(sequenceIDs, column.GetUsesSequenceID(idx))
 	}
-	// TODO(fqazi): Deal with the case where a column is added
-	// in the current statement.
-
+	expressionElem := scpb.DefaultExpression{
+		TableID:         table.GetID(),
+		ColumnID:        column.GetID(),
+		DefaultExpr:     defaultExpr,
+		UsesSequenceIDs: sequenceIDs,
+	}
+	b.addIfDuplicateDoesNotExistForDir(dir, &expressionElem)
 	// Get all available type references and create nodes
 	// for dropping these type references.
-	for _, col := range table.AllColumns() {
-		if !col.HasDefault() || col.ColumnDesc().HasNullDefault() {
-			continue
-		}
-		expr, err := parser.ParseExpr(col.GetDefaultExpr())
+	visitor := &tree.TypeCollectorVisitor{
+		OIDs: make(map[oid.Oid]struct{}),
+	}
+	// Deal with any default expressions with types.
+	expr, err := parser.ParseExpr(defaultExpr)
+	if err != nil {
+		panic(err)
+	}
+	tree.WalkExpr(visitor, expr)
+	for oid := range visitor.OIDs {
+		typeID, err := typedesc.UserDefinedTypeOIDToID(oid)
 		if err != nil {
 			panic(err)
 		}
-		if col.GetID() == id {
-			tree.WalkExpr(visitorDeleted, expr)
-		} else {
-			tree.WalkExpr(visitor, expr)
+		typeRef := &scpb.TypeReference{
+			TypeID: typeID,
+			DescID: table.GetID(),
 		}
-		if col.IsComputed() {
-			expr, err := parser.ParseExpr(col.GetComputeExpr())
-			if err != nil {
-				panic(err)
-			}
-			if col.GetID() == id {
-				tree.WalkExpr(visitorDeleted, expr)
+		b.addIfDuplicateDoesNotExistForDir(dir, typeRef)
+	}
+}
+
+// decomposeDescToElements converts generic parts
+// of a descriptor into an elements in the graph.
+func (b *buildContext) decomposeDescToElements(
+	_ context.Context, table catalog.Descriptor, dir scpb.Target_Direction,
+) {
+	// Decompose all security settings
+	privileges := table.GetPrivileges()
+	ownerElem := scpb.Owner{
+		DescriptorID: table.GetID(),
+		Owner:        privileges.Owner().Normalized(),
+	}
+	b.addIfDuplicateDoesNotExistForDir(dir, &ownerElem)
+
+	for _, user := range privileges.Users {
+		userElem := scpb.UserPrivileges{
+			DescriptorID: table.GetID(),
+			Username:     user.User().Normalized(),
+			Privileges:   user.Privileges,
+		}
+		b.addIfDuplicateDoesNotExistForDir(dir, &userElem)
+	}
+}
+
+func (b *buildContext) decomposeColumnIntoElements(
+	_ context.Context,
+	table catalog.TableDescriptor,
+	column catalog.Column,
+	dir scpb.Target_Direction,
+) {
+	if column.IsHidden() {
+		return
+	}
+	b.addIfDuplicateDoesNotExistForDir(dir,
+		&scpb.ColumnName{
+			TableID:  table.GetID(),
+			ColumnID: column.GetID(),
+			Name:     column.GetName(),
+		},
+	)
+	b.addIfDuplicateDoesNotExistForDir(dir, b.columnDescToElement(table, *column.ColumnDesc(), nil, nil))
```

Comment:
Please use `ColumnDescDeepCopy` instead of `ColumnDesc`, for safety. This comment applies to other usages of `ColumnDesc` or any other `\w+Desc` methods.